File: 01_DFS_BFS/assignment.py
class Graph:
    def __init__(self, num_nodes: int):
        self.num_nodes = num_nodes
        self.adj_list = [[] for _ in range(self.num_nodes)]

    def add_edge(self, node_1: int, node_2: int):
        self.adj_list[node_1].append(node_2)
        self.adj_list[node_2].append(node_1)

    def __str__(self) -> str:
        str_repr = ""
        for i in range(len(self.adj_list)):
            str_repr += f"{i} {self.adj_list[i]}\n"
        return str_repr

    def depth_first_iter(self, key: int):
        stack = [(0, 0)]
        visited = [(0, 0)]
        search_sequence: list[tuple[int, int]] = []
        while len(stack) != 0:
            curr_node = stack[-1]
            curr_node_element = curr_node[0]
            curr_node_level = curr_node[1]
            del stack[-1]
            search_sequence.append(curr_node)
            if curr_node_element == key:
                print(f"Node {curr_node_element} found!")
                break
            neighbors = self.adj_list[curr_node_element]
            for neighbor in neighbors:
                if neighbor not in [visited_element[0] for visited_element in visited]:
                    visited.append((neighbor, curr_node_level + 1))
                    stack.append((neighbor, curr_node_level + 1))
        else:
            print("Node not found by iterative DFS")
        print(f"Search sequence: {search_sequence}")

    def breadth_first_iter(self, key: int):
        queue = [(0, 0)]
        visited = [(0, 0)]
        search_sequence: list[tuple[int, int]] = []
        while len(queue) != 0:
            curr_node = queue[0]
            curr_node_element = curr_node[0]
            curr_node_level = curr_node[1]
            del queue[0]
            search_sequence.append(curr_node)
            if curr_node_element == key:
                print(f"Node {curr_node_element} found!")
                break
            neighbors = self.adj_list[curr_node_element]
            for neighbor in neighbors:
                if neighbor not in [visited_element[0] for visited_element in visited]:
                    visited.append((neighbor, curr_node_level + 1))
                    queue.append((neighbor, curr_node_level + 1))
        else:
            print("Node not found by iterative BFS")
        print(f"Search sequence: {search_sequence}")

    def depth_first_recur(self, key: int):
        visited = []
        self.depth_first_recur_impl(0, visited, key)

    def depth_first_recur_impl(self, node: int, visited: list[int], key: int):
        if node is None:
            return
        if node == key:
            print(f"Visited node {node}")
            print(f"Node {key} found!")
            return
        for neighbor in self.adj_list[node]:
            if neighbor not in visited:
                visited.append(neighbor)
                self.depth_first_recur_impl(neighbor, visited, key)

    def breadth_first_recur(self, key: int):
        queue = [0]
        visited = [0]
        self.breadth_first_recur_impl(queue, visited, key)

    def breadth_first_recur_impl(self, queue: list[int], visited: list[int], key: int):
        if len(queue) == 0:
            return
        front = queue[0]
        del queue[0]
        if front == key:
            print(f"Visited node {front}")
            print(f"Node {key} found!")
            return
        visited.append(front)
        for neighbor in self.adj_list[front]:
            if neighbor not in visited:
                queue.append(neighbor)
        self.breadth_first_recur_impl(queue, visited, key)


graph = Graph(5)
graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(1, 3)
graph.add_edge(1, 4)
print(graph)

graph.breadth_first_iter(3)
graph.depth_first_iter(3)

graph.depth_first_recur(3)
graph.breadth_first_recur(3)

File: 01_DFS_BFS/assignment.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <string>
#include <algorithm>
#include <utility>
using namespace std;

#define LBR '\n'

class Graph {

    map<string,vector<string>> adjacency_list;
    size_t n_nodes ; 
    size_t n_edges ; 

    public:

    void add_edge( 
        const string& node1 , 
        const string& node2 
    ) {
        if (adjacency_list.count(node1) != 0) {
            adjacency_list[node1].push_back( node2 ) ; 
        }
        else {
            vector<string> node_list ; 
            node_list.push_back( node2 ) ; 
            adjacency_list[node1] = node_list;
        }
        if (adjacency_list.count(node2) != 0) {
            adjacency_list[node2].push_back( node1 ) ; 
        }
        else {
            vector<string> node_list ; 
            node_list.push_back( node1 ) ; 
            adjacency_list[node2] = node_list;
        }
    }

    void print() {
        for (const pair<string,vector<string>>& node_list: adjacency_list ) {
            cout << node_list.first << " : " ; 
            for (const string& neighbor: node_list.second ) {
                cout << neighbor << " " ; 
            }
            cout << LBR ; 
        }
    }

    void depth_search(
        const string& key , 
        const string& start_node 
    ) {
        vector<string> visited ; 
        depth_search_impl( start_node , key , visited , 0 ) ;
    }

    bool depth_search_impl( 
        const string& node , 
        const string& key , 
        vector<string>& visited ,
        size_t level
    ) {
        cout << "[DFS] Visited " << node << " , level = " << level << LBR ; 
        if (node == key) {
            cout << key << " found in the graph!" << LBR ; 
            return true;
        }
        visited.push_back( node ) ; 
        vector<string> neighbors = adjacency_list[node] ; 
        for (const string& neighbor: neighbors) {
            if ( find( visited.begin() , visited.end() , neighbor) == visited.end() ) {
                bool found = depth_search_impl( neighbor , key , visited , level + 1 ) ;  
                if (found) return true; 
            }
        }
        return false; 
    }

    void breadth_search(
        const string& key , 
        const string& start_node 
    ) {
        vector<string> visited ; 
        queue<string> frontier ; 
        queue<size_t> frontier_levels ; 
        frontier.push( start_node ) ;
        frontier_levels.push( 0 ) ; 
        breadth_search_impl( frontier , frontier_levels , key , visited ) ;
    }

    void breadth_search_impl( 
        queue<string>& frontier , 
        queue<size_t>& frontier_levels , 
        const string& key , 
        vector<string>& visited 
    ) {
        if (frontier.size() == 0) return ;
        string node = frontier.front() ;
        size_t level = frontier_levels.front() ; 
        cout << "[BFS] Visited " << node << " , level = " << level << LBR ;  
        frontier.pop() ; 
        frontier_levels.pop() ; 
        if (node == key) {
            cout << key << " found in the graph!" << LBR ; 
        }
        visited.push_back( node ) ; 
        vector<string> neighbors = adjacency_list[node] ; 
        for (const string& neighbor: neighbors) {
            if ( find( visited.begin() , visited.end() , neighbor) == visited.end() ) {
                frontier.push( neighbor ) ; 
                frontier_levels.push( level + 1 ) ;
            }
        }
        breadth_search_impl( frontier , frontier_levels , key , visited ) ;  
    }


} ;

string input( string message ) {
    cout << message << LBR ; 
    string line ;
    cin >> line ; 
    return line ;
} ; 

int main( int argc , char* argv[] ) {
    Graph graph ; 
    cout << "Enter node-pairs:" << LBR ; 
    while (true) {
        string node1 = input( "Node 1" ) ; 
        string node2 = input( "Node 2" ) ;  
        if (node1 == "-" && node2 == "-") {
            break ;
        }
        graph.add_edge( node1 , node2 ) ;
        cout << "[Edge from " << node1 << " to " << node2 << " added]" << LBR ; 
    }
    string key = input( "Enter key: " ) ;
    string start_node = input( "Enter start node: " ) ; 
    graph.depth_search( key , start_node ) ; 
    graph.breadth_search( key , start_node ) ; 
    graph.print() ; 
}